### ISSUE TITLE
Linebreaking support

### ISSUE THREAD

**Issue by fred-wang** [w3c/mathml-core#127]

**Comment by NSoiffer** [comment]
Since it is not in this issue and might prove useful to a future core implementation... I implemented a [linebreaking polyfil ](https://github.com/mathml-refresh/mathml-polyfills/tree/main/linebreaking) back in 2020 (seems like a lifetime ago...). You can [see it in action on github.io](https://mathml-refresh.github.io/mathml-polyfills/acid-test.html#linebreaksindisplaymathmlHeader). Click on `Apply Transform` to see it work (if you have Chrome/Edge, the MathML display needs to be on).

This transform makes use of one column `mtable`s as its target because there is currently no other way to get multiple lines to show up in the implementations. If core supported a manual linebreak (i.e., if `<mo linebreak='newline'>` is supported and forced the start of a new line), then this polyfill could take advantage of that and it would be much less intrusive in what it currently does to the MathML by adding an `mtable`.

Note: indentation is done using `mspace` and that same idea would carry forward to a version of core that supported a manual linebreak.

**Comment by dginev** [comment]
Since we have a prolonged gap period here, are there any current recommendations for pure CSS solutions for reflow?

I took a stab at switching the display of a simple equation to `inline-flex` with a corresponding `@media` query for small viewports, and it seemed to behave quite reasonably for a failsafe (in FF and Chrome).

Here is an example of that (with flex always on). The demo should be able to render 6 different arrangements as the screen shrinks:
https://codepen.io/dginev/pen/rNQjdzR

It would take some more fine-tuning to control the finer details of reflow, but this could already be a healthy upgrade for common equation markup.

**Comment by ronkok** [comment]
I can confirm that a flex-based solution works pretty well. You can see it already in action if you navigate to [Temml.org](https://temml.org/#A%20=%20a%20+%20b%20+%20c%20=%20d%20+%20e%20+%20f%20=%20a%20+%20b%20+%20c%20=%20d%20+%20e%20+%20f) and turn display mode off.

In default mode, Temml writes MathML with `<mrow>` elements that each end in a binary operator or relation operator. (Per _The TeXbook_ p. 173) Then the `<math>` element carries the following CSS:

```
/* flex-wrap for line-breaking in Chromium */
math {
  display: inline-flex;
  flex-wrap: wrap;
  align-items: baseline;
}
math > mrow {
  padding: 0.5ex 0ex;
}

/* Avoid flex-wrap in Firefox */
@supports (-moz-appearance:meterbar) and (display:flex) {
  math { display: inline; }
  math > mrow { padding: 0 }
}
```

I don’t apply flex-wrap in Firefox. In Firefox, the separation by `<mrow>` elements already works without a flexbox.

Temml has a rendering option which allows a website administrator to select breaks before `=` signs instead of after binary operators. The CSS remains the same, but Temml generates MathML with differently grouped `<mrow>` elements.

In the comments above, I was one of those asking for line-breaking action in the MathML Core specification. At this time, I do not make that request. In Chromium and Firefox, I think line-breaking is largely a solved problem. Chromium and WebKit have much bigger rendering problems to solve.

Sadly, flex-based line-breaking does not work in WebKit. Maybe someday.

**Comment by dginev** [comment]
@ronkok This is great, thank you for the extra context.

But I wouldn't go beyond calling this a "stopgap solution", since you've enumerated some serious problems. WebKit lacking support is one, another is ending up with two non-standard solutions for Chrome and Firefox. 

Not having a standard way to manually force (or softly suggest) a linebreak in the usual MathML markup is a third. Flexbox allows a variety of techniques to support what Neil referred to as `<mo linebreak='newline'>`, but none seems to work on an element with pre-set content. Instead, with flexbox we seem to need a dedicated empty element to indicate the forced break (similar to the empty `<br>` in HTML). Here is an example of the best I could come up with. Since it had to be empty, I used `<mspace>`):
https://codepen.io/dginev/pen/zYMNjgd

In summary:

```html
<mrow> ...LHS... </mrow>
<mo>=</mo>
<mspace class="linebreak"></mspace>
<mrow> ...RHS... </mrow>
```

```css
mspace.linebreak {
  flex-basis: 100%;
  height: 0;
  width: 0;
  overflow: hidden;
}
```

As an alternative (and maybe even worse) trick, one could rearrange the `<mrow>` structure by tucking in the equal sign in the left-hand side mrow, which will have them treated as a single flex item. But that would break the default spacing support available via the operator dictionary. So I think inline-flex "mostly working" is a nice surprise, but still appears to be a crutch to ducktape some reflow together, until we have a proper mechanism.

**Comment by bkardell** [comment]
> In Chromium and Firefox, I think line-breaking is largely a solved problem. Chromium and WebKit have much bigger rendering problems to solve.

One of these is a typo, I suppose? The second one?

**Comment by ronkok** [comment]
To clarify: Temml's `<mrow>` trick works pretty well at providing non-display mode line-breaking in both Chromium and Firefox. It does not work in WebKit.

On the more general question, I have begun compiling a list of [browser issues](https://temml.org/docs/en/administration.html#browser-issues). Firefox is the best, by a large margin. Chromium has some serious issues, especially if system fonts are used. WebKit is the worst. It cannot even render an accent at the correct vertical alignment. 

I put a lot of work into the Temml library. It is my sincere hope that someday, MathML will get widespread use. But I think that day is not yet.  I suspect most web site administrators will avoid MathML until browser rendering is more reliable.

**Comment by NSoiffer** [comment]
@ronkok: I'm a little late to the game, but I finally tried temml.org in Chrome and it is great to see flex-based solution works. But putting aside it currently only works in Chrome/Edge, I would add it is only half the solution. Not only does an expression need to wrap long lines, it needs to _indent_ them appropriately. Is that something that can be done with flexbox?

The [polyfill I mentioned](https://mathml-refresh.github.io/mathml-polyfills/acid-test.html#linebreaksindisplaymathmlHeader) illustrates why that's important.  But it is hacky in that it had to create an mtable. I hope core level 2 will have at least the hook mentioned.

**Comment by ronkok** [comment]
@NSoiffer I'd like to look at your polyfill in action, but I'm getting a 404 error when I click on that link.

I'm still thinking about how to get an indent. I have a version working in a web app I call [Hurmet.app](https://hurmet.app/sample). It renders math with Temml's version of MathML and I have it set to wrap before top-level `=` characters. Here is a screenshot of how that looks:
![Math wrap](https://github.com/w3c/mathml-core/assets/16403058/eb5e6b13-4268-49b1-acd8-8e2afb5cb063)

To get that indent, I have appended the following CSS.
```
/* Create a hanging indent on calculations that wrap to a second line. */
.hurmet-calc > math > mrow:not(:first-child) { margin-left: 2em }
.hurmet-calc > math > mrow:not(:last-child) { margin-right: -2em }
```

That's pretty hacky and I have applied it only to my own site, Hurmet, not to the library Temml. I'd like better control over the width of the indent. I'd like something less odd.  But it will serve as a temporary line breaking solution. I agree with both you and @dginev that this is a temporary fix and would benefit from better support in MathML-Core someday.

But there several other rendering issues that I think should get higher priority.

**Comment by davidcarlisle** [comment]
@NSoiffer  @ronkok  fixed at
https://w3c.github.io/mathml-polyfills/acid-test.html

**Comment by ronkok** [comment]
Thank you for the link to the polyfill. It's nice work.

Temml is written to run either client-side or server-side. It therefore does not have access to `document.getElementById()` and cannot use the techniques in the polyfill. Temml line-breaking is a CSS solution.

There is room in the world for both CSS solutions and JavaScript solutions. Hopefully, one day the browser will have a native solution and neither will be necessary.
